### PR TITLE
Implemented C++ Message Generation

### PR DESCRIPTION
**Pull request by garethellis0**
The two significant changes here are:
- Implemented message generation for C++ in `message_generation.bzl`. I tried to be as similar as possible to the python message generation.
- Added python pip dependency management via Bazel. The impetus is that a requirement was missing (`empy`), and I thought this was a cleaner way of getting it then installing all the dependencies manually.

I have some changes planned after this to allow this to be used with other bazel projects without the need to run the `compile_ros_workspace.sh` rule. I think this will really help make this more accessible for others, would love to have any immediate feedback you might have.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -16,8 +16,7 @@ rosdistro==0.6.2
 rosinstall-generator==0.1.13
 rospkg==1.1.1
 six==1.10.0
-trollius==2.1
+trollius==2.1.post2
 urllib3==1.7.1
 vcstools==0.1.39
-wsgiref==0.1.2
```
Comment (garethellis0) [inline requirements.txt]: `wsgiref` is part of the standard library in python3. Bazel uses python3, not 2, so the pypi dependency management broke with this (the pypi `wsgiref` is not python3 compatible). Please feel free to let me know if this is needed for python2 and I can find a workaround.

File: requirements.txt [hunk 2]
```diff
@@ -16,8 +16,7 @@ rosdistro==0.6.2
 rosinstall-generator==0.1.13
 rospkg==1.1.1
 six==1.10.0
-trollius==2.1
+trollius==2.1.post2
```
Comment (garethellis0) [inline requirements.txt]: Pip for python3 could only find `2.1.post2` when running the python dependency management via bazel. This library is not technically supported for python 3, so frankly I'd like to remove it, but I left it in in the interests of minimizing changes.

**Comment by nicolov** [comment]
> Added python pip dependency management via Bazel.

Thanks for this. Encouraged by your interest, I just merged a PR to remove pip altogether instead. I think this is better because it leaves people free to do whatever they want to manage Python stuff in Bazel (which is kind of a pain). Can you update your PR to match it?

> I tried to be as similar as possible to the python message generation.

Lol, but that is kind of garbage. I don't think it properly supports transitive message dependencies at all (meaning you won't be able to depend on Bazel-generated messages), and what about the random target name :D In any case, let's merge this once you remove all the Python stuff. If you're determined about getting this to work, I can guide you through:

- removing Bazel-managed Python libs and just using the message generator installed by catkin, which works just fine and is easier to maintain

- doing transitive messages and target naming properly

> allow this to be used with other bazel projects without the need to run the compile_ros_workspace.sh rule

Please note that I added the compilation step because I knew I had to support patching the roscpp code. If you don't need that, the debian packages work just fine.

In general, keep in mind that is a Frankenstein approach - mixing catkin and bazel things in the same project. I do think it has merit in that it allows one to get started without having to maintain Bazel definitions for all ROS packages. However, in the long run (and for large teams) it's probably not the best approach. At Cruise, I ended up rebuilding everything with Bazel to support easy patching, sanitizers, compiler options, etc..

**Comment by garethellis0** [comment]
Apologies for taking so long to get to back to this, exams and the holiday season had to take priority for a while. Glad to hear that you're open to improving this! I definitely noticed some funky things in the message generation, but thought it best to keep the changes minimal to keep review easy :sweat_smile:. 

I've removed all the pypi stuff, and confirmed the tests for c++ message generation still pass.

Edit: unless `empy` is installed on the system python3, looks like bazel will fail because it doesn't use the local virtualenv. So we might need to skip right to using message generator installed with catkin in this PR, unless you're happy with me adding `sudo apt-get install python3-empy` to CI?

**Comment by garethellis0** [comment]
> * removing Bazel-managed Python libs and just using the message generator installed by catkin, which works just fine and is easier to maintain
> 
> * doing transitive messages and target naming properly

I think I've got a pretty good idea of how to get transitive messages and target naming working, but I'm less clear on using the message generator installed by catkin. Could you elaborate?

**Comment by nicolov** [comment]
Do you mind if I take a crack at this over the next week?

**Comment by garethellis0** [comment]
Sure, no problem. Thanks!

**Comment by garethellis0** [comment]
Hey @nicolov - just following up on this, any progress or anything I can help out with?

**Comment by nicolov** [comment]
> Hey @nicolov - just following up on this, any progress or anything I can help out with?

You know, I looked at the situation and it's not that bad, so I'm going to go ahead and merge it. The main thing this is missing right now is the ability to depend on messages generated from other Bazel packages (ie a `deps` attribute for the `generate_messages` rule). Is that something you might need in the future?

**Comment by garethellis0** [comment]
Sorry, just seeing this comment now. I am going to need transitive messages. Totally happy to implement that, please let me know if there's a particular direction it you were thinking of, otherwise I'm sure I can figure something out.

**Comment by nicolov** [comment]
I’ve done that, feel free to try the code in master.

**Pull request merged**